txjuju/config.py

```diff
@@ -0,0 +1,130 @@
+
+
+class Config(object):
+    """An encapsulation of Juju's local configuration."""
+
+    def __init__(self, *controllers):
+        """
+        @param controllers: The ControllerConfig for each of the
+            controllers in the local configuration.
+        """
+        self._controllers = controllers
+
+
+class ControllerConfig(
+        namedtuple("ControllerConfig", "name cloud bootstrap")):
+    """An encapsulation of the local configuration for a single controller."""
+
+    DEFAULT_SERIES = "trusty"
+
+    @classmethod
+    def from_info(cls, name, type,
+                  cloud_name=None, default_series=None, admin_secret=None):
+        """Return a new controller config for the given info.
+
+        @param name: The name of the controller.
+        @param type: The controller's provider type.
+        @param cloud_name: The name of the cloud (defaults to <name>-<type>).
+        @param default_series: The OS series to provision by default.
+            If not provided, it defaults to trusty.
+        @param admin_secret: The password to use for the admin user,
+            if any.
+        """
+        if cloud_name is None and name and type:
+            cloud_name = "{}-{}".format(name, type)
+        # TODO Sort these out as soon as we need them.
+        endpoint = auth_types = credentials = None
+        return cls(
+            name,
+            CloudConfig(cloud_name, type, endpoint, auth_types, credentials),
+            BootstrapConfig(default_series, admin_secret),
+            )
+
+    def __new__(cls, name, cloud, bootstrap=None):
+        """
+        @param name: The name of the controller.
+        @param cloud: The controller's cloud config.
+        @param bootstrap: The controller's bootstrap config, if any.
+        """
+        name = unicode(name) if name else None
+        cloud = CloudConfig._make(cloud) if cloud else None
+        if bootstrap is None:
+            bootstrap = BootstrapConfig("")
+        else:
+            bootstrap = BootstrapConfig._make(bootstrap)
+        return super(ControllerConfig, cls).__new__(
+            cls, name, cloud, bootstrap)
+
+    def __init__(self, *args, **kwargs):
+        super(ControllerConfig, self).__init__(*args, **kwargs)
+        if not self.name:
+            raise ValueError("missing name")
+        if not self.cloud:
+            raise ValueError("missing cloud")
+        if not self.bootstrap:
+            raise ValueError("missing bootstrap")
+
+
+class CloudConfig(
+        namedtuple("CloudConfig",
+                   "name type endpoint auth_types credentials")):
+    """An encapsulation of the local config for a single cloud."""
+
+    def __new__(cls, name, type=None, endpoint=None,
+                auth_types=None, credentials=None):
+        """
+        @param name: The cloud's user-defined ID.
+        @param type: The provider type (defaults to the name).
+        @param endpoint: The endpoint to use, if needed.
```

Comment:
can we add simple examples in the docstring of expected formats for these parameters.
For example, should endpoint begin w/ http://?
Shall we mention type should be one of lxd, maas, openstack etc?